### PR TITLE
implement achievement system

### PR DESCRIPTION
**Pull request by Henrybk**
This is a pull request to organize the new achievement system work.

I am using this pull request in Hercules WS as the base: https://github.com/HerculesWS/Hercules/pull/1666/

### CODE REVIEW

File: src/Network/Receive/iRO.pm [hunk 1]
```diff
@@ -42,6 +45,92 @@ sub new {
...
+} __attribute__((packed));
+=cut
+
+sub achievement_list {
+	my ($self, $args) = @_;
+	
+	my $msg = $args->{RAW_MSG};
+	my $msg_size = $args->{RAW_MSG_SIZE};
+	my $headerlen = 22;
+	my $achive_len = 50;
```
Comment (allanon) [inline src/Network/Receive/iRO.pm]: Spelling. `$achieve_len`

Comment (allanon) [inline src/Network/Receive/iRO.pm]: I like doing something like:

```
my $achieve_pack = 'V C V10 V C';
my $achieve_len = pack $achieve_pack;
```

... for less magic variables. What do you think?

**Comment by Henrybk** [comment]
The new send function is working, tested it with the console command `eval $messageSender->sendAchievementGetReward(240001)` and got `Increase AGI` status for 60s.

File: src/Network/Receive/iRO.pm [hunk 2]
```diff
@@ -33,6 +33,9 @@ sub new {
 		'0097' => ['private_message', 'v Z24 V Z*', [qw(len privMsgUser flag privMsg)]], # -1
 		'082D' => ['received_characters_info', 'x2 C5 x20', [qw(normal_slot premium_slot billing_slot producible_slot valid_slot)]],
 		'099D' => ['received_characters', 'x2 a*', [qw(charInfo)]],
+		'0A23' => ['achievement_list', 'v v V V v V V', [qw(len ach_count total_points rank current_rank_points next_rank_points)]], # -1
```
Comment (allanon) [inline src/Network/Receive/iRO.pm]: This seems wrong. The number of values unpacked (7) doesn't match the number of keys (6).

Comment (Henrybk) [inline src/Network/Receive/iRO.pm]: Thanks

File: src/Network/Receive/iRO.pm (hunk 1 above)
Comment (kaliwanagan) [inline src/Network/Receive/iRO.pm]: I like the length better too :)

File: src/Commands.pm [hunk 3]
```diff
@@ -56,6 +56,7 @@ our %customCommands;
 sub initHandlers {
 	%handlers = (
 	a					=> \&cmdAttack,
+	achievegetreward	=> \&cmdAchieveGetReward,
```
Comment (allanon) [inline src/Commands.pm]: Can we please use a smaller command name for this? Maybe `achieve` or `ach` as the base command, then `achieve list`, `achieve reward ID`, etc.

File: src/Commands.pm [hunk 4]
```diff
@@ -6098,4 +6099,47 @@ sub cmdDeadTime {
...
+				"eventMacro include on all\n".
+				"eventMacro include off <filename or pattern>\n".
+				"eventMacro include off all\n".
+				"eventMacro include list\n", 'list';
+		return;
+	}
+
+	if ($arg1 eq 'reward') {
+		if (!exists $achievementList->{$arg2}) {
+			error "You don't have the achievement $arg2.\n";
```
Comment (kaliwanagan) [inline src/Commands.pm]: No translations? :D 🗡 
(lol I know this has already been long time under review :P)

@allanon are we cool with merging this? we can always improve it again later on (as long as it doesn't break backwards compatibility)

Comment (allanon) [inline src/Commands.pm]: Sounds good to me!

Comment (sctnightcore) [inline src/Commands.pm]: <3

**Comment by ya4ept** [comment]
we forgot to add a command to the documentation

**Pull request merged**